### ISSUE TITLE
milkytracker: Event port write failed: bad port ID

### ISSUE THREAD

**Issue by danboid** [haikuports/haikuports#6488]
milkytracker crashes on launch.

[MilkyTracker-1106-debug-21-12-2021-23-41-51.report.zip](https://github.com/haikuports/haikuports/files/7759047/MilkyTracker-1106-debug-21-12-2021-23-41-51.report.zip)

**Comment by Begasus** [comment]
Seems to be ok on 32bit (version 1.03.00)

**Comment by danboid** [comment]
We had this conversation on IRC about this earlier:

<diver> ok, it seems milkytracker build is broken 
<diver> execution_defs.h: No such file or directory
* Begasus_64 (~Begasus_6@d54C05662.access.telenet.be) has joined
<danboid> diver, Ah! Thanks for trying
<danboid> diver, Do you know why thats missing now?
<diver> nope
<Begasus> k, just checked a 64bit version, crashes here also
<diver>  /boot/system/develop/tools/lib/gcc/x86_64-unknown-haiku/11.2.0/include/c++/pstl/glue_algorithm_defs.h:15:10: fatal error: execution_defs.h: No such file or directory
<Begasus> diver, are you on 64bit?
<diver> yes
<diver> crashes here too
<Begasus> maybe the update on gcc trigered something here
* HaikuUser (~vision@dynamic-077-000-226-058.77.0.pool.telefonica.de) has joined
* HaikuUser has quit ()
<Begasus> https://github.com/milkytracker/MilkyTracker/search?q=defs.h
<diver> ok, have an idea
* freakazoid343 (~matt@2603:9000:cf0f:80e3:600b:877:2a65:889e) has joined
<Begasus> searching for defs.h doesn't reveal anyting there
<diver> it's in gcc_syslibs_devel
* freakazoid12345 has quit (Ping timeout: 480 seconds)
<Begasus> checking
* freakazoid12345 (~matt@2603-9000-cf0f-80e3-600b-0877-2a65-889e.inf6.spectrum.com) has joined
<diver> interesting, the header is there in chroot but gcc doesn't see it for some reason
<Begasus> searching the system reveals "/boot/system/develop/headers/private/kernel/arch/sh4/" ?
* freakazoid343 has quit (Ping timeout: 480 seconds)
<Begasus_64>  /boot/system/develop/tools/lib/gcc/x86_64-unknown-haiku/11.2.0/include/c++/pstl/glue_algorithm_defs.h:15:10: fatal error: execution_defs.h: No such file or directory
<Begasus> another header ...
<Begasus> also included in gcc_syslibs_devel
<diver> building outside of chroot also doesn't work

**Comment by diversys** [comment]
Seems like to fix the build we need to include `/system/develop/headers/c++/pstl` in Jamrules? 
This is where `execution_defs.h` is located.

**Comment by Begasus** [comment]
@diversys maybe create a ticket for it at the tracker? It will get a better chance to be included then mentioning it here? ;)

**Comment by diversys** [comment]
I think this ticket is fine for now.

**Comment by kenmays** [comment]
See: https://github.com/haikuports/haikuports/blob/master/media-sound/milkytracker/milkytracker-1.03.00.recipe

The GCC > 2 was known to crash per the recipe. Also, the x86-GCC_11.2 build fails at the same point as the x64_GCC11.2 build.
Only the x86-GCC_2 build is known to work at the moment.

**Comment by diversys** [comment]
Forcing std=c++11 fixes the build, however the crash is still there. App doesn't crash unless you move a mouse over its window.
```
diff --git a/platforms/haiku/Jamfiles/Jamrules b/platforms/haiku/Jamfiles/Jamrules
index dfce893..98c4725 100644
--- a/platforms/haiku/Jamfiles/Jamrules
+++ b/platforms/haiku/Jamfiles/Jamrules
@@ -1,5 +1,5 @@
 C++ = g++ ;
-C++FLAGS = -O2 -DMILKYTRACKER -D__HAIKU__ ;
+C++FLAGS = -O2 -DMILKYTRACKER -D__HAIKU__ -std=c++11 ;
```

**Comment by diversys** [comment]
Recompiling with '-g" made this crash go away.

**Comment by danboid** [comment]
> Recompiling with '-g" made this crash go away.

This fixes Milkytrackers crash with the latest gcc and x86_64 does it?

Doesn't -g just enable (more) debugging support? That doesn't sound like the best fix.

**Comment by pulkomandy** [comment]
Yes that's not really a fix. But it makes things harder to debug if building in debug mode results in no crash.

**Comment by danboid** [comment]
I didn't actually get round to trying to build this port until yesterday.

The port still compains about execution_defs.h being missing when it tries to build milkytracker but I can see the file is installed at:

/boot/system/develop/tools/lib/gcc/x86_64-unknown-haiku/11.2.0/include/c++/pstl/execution_defs.h

How do I modify the recipe / jamrules file or whatever to include this path and not have it break every time g++ gets updated to a new version?

**Comment by danboid** [comment]
Milkytracker gets as far as trying to build SampleEditorControl.cpp, line 35 of which is:

#include <algorithm>

I can see cc1plus is then pulling in

/boot/system/develop/tools/lib/gcc/x86_64-unknown-haiku/11.2.0/include/c++/pstl/glue_algorithm_defs.h before complaining that execution_defs.h doesn’t exist.

I am unable to edit /boot/system/develop/tools/lib/gcc/x86_64-unknown-haiku/11.2.0/include/c++/pstl/glue_algorithm_defs.h to attempt to insert the extra pstl before execution_defs.h, I presume because its within a package?

Looking at the header/comments at the top of glue_algorithm_defs.h, it seems to be an include file from LLVM but haikuporter seems to be using gcc to build it. Can I safely use LLVM header files with gcc? It seems that isn’t the case here.

**Comment by pulkomandy** [comment]
Headers stored in gcc/x86_64-unknown-haiku/11.2.0 are clearly coming from GCC 11, what makes you think they are from LLVM?

> Forcing std=c++11 fixes the build

Seems the correct fix to me. Probably the code is not compatible with newer gcc versions. In fact the cross-platform CMakeLists forces use of c++98 (an even older standard): https://github.com/milkytracker/MilkyTracker/blob/master/CMakeLists.txt#L26

I am not sure why the Haiku port use a Jamfile instead of using CMake, it would avoid confusion.

**Comment by korli** [comment]
fixed in 8be7a61d844d9d0b4e406c4c0ec7505047fa4cfb

**Comment by danboid** [comment]
Thanks korli!
